File: cpp-std/demo11b-exec-service.cpp
/*
EXECUTOR SERVICES AND THREAD POOLS

Executor services in C++ std threading are not supported by default.
So, I use mylib::ExecService for this demonstration.
*/


#include <iostream>
#include <chrono>
#include "mylib-exec-service.hpp"
using namespace std;



int main() {
    constexpr int NUM_THREADS = 2;
    constexpr int NUM_TASKS = 5;

    auto execService = mylib::ExecService();
    execService.init(NUM_THREADS);

    for (int i = 0; i < NUM_TASKS; ++i) {
        execService.submit([=] {
            char id = 'A' + i;
            std::cout << "Task " << id << " is starting" << std::endl;
            std::this_thread::sleep_for(std::chrono::seconds(3));
            std::cout << "Task " << id << " is completed" << std::endl;
        });
    }

    std::cout << "All tasks are submitted" << std::endl;

    execService.waitTaskDone();
    std::cout << "All tasks are completed" << std::endl;

    execService.shutdown();

    return 0;
}

File: python/demo11b_exec_service.py
'''
EXECUTOR SERVICES AND THREAD POOLS
Version B: The executor service containing multiple threads
'''

import time
from concurrent.futures import ThreadPoolExecutor



def do_task(name: str):
    print(f'Task {name} is starting')
    time.sleep(3)
    print(f'Task {name} is completed')



NUM_TASKS = 5
executor = ThreadPoolExecutor(max_workers=2)

for i in range(NUM_TASKS):
    task_name = chr(i + 65)
    executor.submit(do_task, task_name)

executor.shutdown(wait=True)
